LFS-744: Questionnaire editor does not auto-create minAnswers and maxAnswer fields

veronikaslc opened the pull request.

Commit:
LFS-744: Questionnaire editor does not auto-create minAnswers and maxAnswer fields

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/NumberInput.jsx:
```diff
@@ -34,16 +34,26 @@ import QuestionComponentManager from "./QuestionComponentManager";
 
 let NumberInput = (props) => {
   let { objectKey, data } = props;
+  let defaultToZero = ["minAnswers", "maxAnswers"].includes(objectKey) || objectKey.includes('maxPerSubject');
```
Not a fan of this at all. Hardcoding `maxPerSubject` was already pretty bad.

Looking at our JSONs that define the properties for questionnaire, section, and question, I would say it's currently safe to default any property of type `long` to zero.

marta- replied:
Additionally, this fix does not cover the cases of Boolean and Pedigree type questions, where minAnswers is displayed as a dropdown (to restrict the values to 0 or 1) instead of a number field.

Commit:
LFS-744: Questionnaire editor does not auto-create minAnswers and maxAnswer fields
Default types long input to 0

sashaandjic commented:
Problem found. I created a new question and saved it.
![image](https://user-images.githubusercontent.com/23243244/110074861-e6066800-7d4f-11eb-8cfc-e52a12f83f9d.png)

When I opened it int the edit mode again, Data type was gone:
![image](https://user-images.githubusercontent.com/23243244/110074935-ffa7af80-7d4f-11eb-87bb-460e08597d6d.png)

veronikaslc commented:
@sashaandjic did you mean max and min answers were gone?

Commit:
LFS-744: Questionnaire editor does not auto-create minAnswers and maxAnswer fields
Adressed codereview comments

sashaandjic reviewed (approved):
tested

marta- commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/ObjectInput.jsx:
```diff
@@ -36,15 +36,16 @@ import QuestionComponentManager from "../questionnaireEditor/QuestionComponentMa
 let ObjectInput = (props) => {
   let { objectKey, value, data } = props;
-  let [ selectedValue, setSelectedValue ] = useState(data[objectKey] || '');
+  const defaultValue = data[objectKey] || (objectKey === "minAnswers" ? 0 : '');
```
I'm not a big fan of hardcoding `minAnswers` here. How does everyone feel about auto-selecting the first value in the dropdown by default, would there be any unwanted side effects? @sdumitriu @veronikaslc 
```suggestion
  const defaultValue = data[objectKey] || (Object.keys(value || {})[0] || '');
```
If the above change is made, it will of course affect all nested definitions such as `displayMode` and `dataType`.  We should also reorder keys in Question.json so that the default would be more common in practice, e.g:
```
     "displayMode": {
       "input": {},
       "list": {...},
       "list+input" : {...},
       "textbox": {},
       "hidden": {},
       "none": {}
     }
```
as well as putting `text` at the top of the `dataType` list.

veronikaslc replied:
Addressed in 696870a

Commit:
LFS-744: Questionnaire editor does not auto-create minAnswers and maxAnswer fields
Made default to all selections

sashaandjic reviewed (approved):
tested

sdumitriu reviewed (changes requested):
Critical bug: editing the questionnaire properties causes the questionnaire to be almost impossible to create, since `maxPerSubject` becomes 0, and all subjects already have 0 forms of that type.

- go to Administration / Questionnaires / Radiotherapy
- edit questionnaire properties
- enter a description, save
- go to dashboard
- create a subject
- open the subject
- try to create a new Radiotherapy form for it -> not allowed
- It is allowed to create a new Radiotherapy form from the dashboard by creating a new form + a new subject for it
- It is impossible to remove the `0` setting for `maxPerSubject`
- [x] We should allow `0` as a special value for `maxPerSubject` meaning `no limit` (LFS-1034)
- [x] In the questionnaire editor we currently show `Unlimited` if no value is set, but `0` if 0 is set. Should we treat 0 as a special value and also display it as unlimited in view mode?

marta- committed:
LFS-1034: Setting `maxPerSubject=0` should allow unlimited forms for each subject
Quick fix

marta- committed:
LFS-744: Questionnaire editor does not auto-create minAnswers and maxAnswer fields
[Addressing review comments] Fixed UI regression: when maxPerSubject is 0, display 'Unlimited' in view mode

marta- commented:
> * [x]  We should allow `0` as a special value for `maxPerSubject` meaning `no limit` (LFS-1034)

Addressed in b97a410 .

> * [x]  In the questionnaire editor we currently show `Unlimited` if no value is set, but `0` if 0 is set. Should we treat 0 as a special value and also display it as unlimited in view mode?

Done in 9fdf8c4 .

marta- committed:
LFS-1034: Setting `maxPerSubject=0` should allow unlimited forms for each subject
Sometimes maxPerSubject is created as String; ensure it's always treated as a number.

sdumitriu commented on modules/data-entry/src/main/frontend/src/questionnaireEditor/NumberInput.jsx:
```diff
@@ -34,16 +34,25 @@ import QuestionComponentManager from "./QuestionComponentManager";
-        min={objectKey.includes('maxPerSubject') ? 0 : ''}
+        value={value}
+        onChange={(event) => { setValue(event.target.value); }}
+        onBlur={(event) => { setValue(event.target.value || defaultValue); }}
```
I fear that further down the road this may lead to impossible values, like being unable to set "0" for a field that has a positive default value, or "false" if the default is "true".

sashaandjic commented:
retested and approved before the merge.
